virttest/virsh.py

```diff
@@ -621,6 +624,93 @@ def kosher_args(remote_ip, uri):
         return True not in all_false
 
 
+class EventNotFoundError(Exception):
+    """
+    Error when certain event cannot be found.
+    """
+
+    def __init__(self, details=''):
+        self.details = details
+        Exception.__init__(self)
+
+    def __str__(self):
+        return str(self.details)
+
+
+class EventTracker(object):
+
+    @staticmethod
+    def start_get_event(vm_name):
+        """
+        Use a virsh session with subcommand 'event' to catch events
+
+        :param vm_name: name of the vm to be catched
+        :return: the virsh session with 'event'
+        """
+        virsh_session = aexpect.ShellSession(VIRSH_EXEC)
+        event_cmd = 'event %s --all --loop' % vm_name
+        logging.info('Sending "%s" to virsh shell', event_cmd)
+        virsh_session.sendline(event_cmd)
+
+        return virsh_session
+
+    @staticmethod
+    def finish_get_event(virsh_session):
+        """
+        Stop virsh session and return the event output
+
+        :param virsh_session: virsh session to catch events
+        :return: actual event output catched
+        """
+        virsh_session.send_ctrl("^C")
+        time.sleep(5)
+        event_output = virsh_session.get_stripped_output()
+        virsh_session.close()
+        logging.debug('Event output is %s:', event_output)
+
+        return event_output
+
+    @staticmethod
+    def wait_event(func):
+        """
+        Decorator aiming to return until certain type of virsh event defined
+        by func was found
+        """
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+
+            def _get_arg_value(arg):
+                return kwargs.get(arg) if arg in kwargs else \
+                    inspect.signature(func).parameters[arg].default
+
+            def _get_event_output(session):
+                output = session.get_stripped_output()
+                logging.debug(output)
+                return output
+
+            wait_for_event = _get_arg_value('wait_for_event')
+            event_type = _get_arg_value('event_type')
+            event_timeout = _get_arg_value('event_timeout')
+
+            if wait_for_event is True and event_type is not None:
+                virsh_session = EventTracker.start_get_event(str(args[0]))
+                ret = func(*args, **kwargs)
```

Comment:
ret=func(*args, **kwargs) need always executed ,so pull level up before if wait_for_event.
In this way, you can have consistent return :ret

Reply:
@chunfuwen way ahead of you....
Please see line 709. Maybe you need to wait a little bit longer for the page to fully loaded. Because github is a little bit slow recently.